### PR TITLE
CHIP-SPV Integration

### PR DESCRIPTION
**Pull request by pvelesko**
* more consistent use of includes <..>/"...", reserve <> for system includes and "" for others

### CODE REVIEW

**Comment by pvelesko** [comment]
@gargrahul I closed the other branch and opened this. This branch is on my personal fork so that any of you can push changes.

File: include/hip/hip_runtime_api.h [hunk 1]
```diff
@@ -6965,11 +6989,12 @@ static inline hipError_t hipMallocManaged(T** devPtr, size_t size,
...
 }
 
-#endif
-#endif
+#endif // __cplusplus
 
-#include <hip/amd_detail/amd_hip_runtime_pt_api.h>
+#if (defined(__HIP_PLATFORM_HCC__) || defined(__HIP_PLATFORM_AMD__)) && !(defined(__HIP_PLATFORM_NVCC__) || defined(__HIP_PLATFORM_NVIDIA__))
+#include "hip/amd_detail/amd_hip_runtime_pt_api.h"
+#endif // TODO
```
Comment (pvelesko) [inline include/hip/hip_runtime_api.h:6997]: Why have this as a separate file?

File: tests/catch/CMakeLists.txt [hunk 2]
```diff
@@ -49,9 +53,10 @@ if(UNIX)
     set(CMAKE_CXX_COMPILER "${HIP_PATH}/bin/hipcc")
     set(CMAKE_C_COMPILER "${HIP_PATH}/bin/hipcc")
     set(HIPCONFIG_EXECUTABLE "${HIP_PATH}/bin/hipconfig")
-    execute_process(COMMAND perl ${HIPCONFIG_EXECUTABLE} --version
+    execute_process(COMMAND ${HIPCONFIG_EXECUTABLE} --version
```
Comment (pvelesko) [inline tests/catch/CMakeLists.txt]: you don't need to invoke the perl command explicitly as far as I know. This should allow for using both HIPCC and Perl based hipcc

File: include/hip/hip_common.h [hunk 3]
```diff
@@ -54,9 +53,13 @@ THE SOFTWARE.
...
-    (defined(__CUDA_ARCH__) && __CUDA_ARCH__ != 0)
+// // Auto enable __HIP_DEVICE_COMPILE__ if compiled in HCC or NVCC device path
+// #if (defined(__HCC_ACCELERATOR__) && __HCC_ACCELERATOR__ != 0) || \
+//     (defined(__CUDA_ARCH__) && __CUDA_ARCH__ != 0)
+// #define __HIP_DEVICE_COMPILE__ 1
+// #endif
+// TODO
+#if (defined(__clang__) && defined(__HIP_DEVICE_COMPILE__))
 #define __HIP_DEVICE_COMPILE__ 1
 #endif
```
Comment (pvelesko) [inline include/hip/hip_common.h:65]: I'm confused about this entire section

File: tests/catch/CMakeLists.txt [hunk 4]
```diff
@@ -110,6 +115,8 @@ include_directories(
     "./kernels"
     ${HIP_PATH}/include
     ${JSON_PARSER}
+    ${CHIP_DIR_}/HIP/include  #TODO
+    ${CHIP_DIR_}/include # TODO
```
Comment (pvelesko) [inline tests/catch/CMakeLists.txt]: WIP

**Comment by pvelesko** [comment]
With this branch I am able to build CHIP-SPV and the samples. Currently working on unit tests. CHIP-SPV currently lacks ability to call `free()` and `malloc()` from inside device code:

```
/home/pvelesko/space/CHIP-SPV/HIP/tests/catch/multiproc/deviceAllocationMproc.cc:45:47: error: no matching function for call to 'malloc'
    dev_common_ptr = reinterpret_cast<char*> (malloc(size));
                                              ^~~~~~
/usr/include/stdlib.h:539:14: note: candidate function not viable: call to __host__ function from __global__ function
extern void *malloc (size_t __size) __THROW __attribute_malloc__
             ^
/home/pvelesko/space/CHIP-SPV/HIP/tests/catch/multiproc/deviceAllocationMproc.cc:83:7: error: no matching function for call to 'free'
      free(dev_common_ptr);
      ^~~~
/usr/include/stdlib.h:565:13: note: candidate function not viable: call to __host__ function from __global__ function
extern void free (void *__ptr) __THROW;
```

Disabling the build of this test for now if using CHIP-SPV.

**Comment by pvelesko** [comment]
Encountered the following error while building tests: @gargrahul 
```
[ 67%] Building CXX object catch/catch_tests/unit/memory/CMakeFiles/MemoryTest.dir/hipHostMalloc.cc.o
In file included from /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/unit/memory/hipHostMalloc.cc:32:
In file included from /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_checkers.hh:24:
/gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:242:38: error: no matching function for call to 'getExpectedArgs'
  using expectedArgsTuple = decltype(getExpectedArgs(f));
                                     ^~~~~~~~~~~~~~~
/gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:267:3: note: in instantiation of function template specialization 'HipTest::validateArguments<void (*)(int *, int), int *, int>' requested here
  validateArguments(kernel, packedArgs...);
  ^
/gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/unit/memory/hipHostMalloc.cc:69:14: note: in instantiation of function template specialization 'HipTest::launchKernel<void (*)(int *, int), dim3, int *&, int>' requested here
    HipTest::launchKernel(Set, dimGrid, dimBlock, 0, 0x0, ptr, -42);
             ^
/gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:228:51: note: candidate template ignored: deduced type 'void (*)(int *, int)' of 1st parameter does not match adjusted type 'void (*)(int *, int)' of argument [with FArgs = <int *, int>]
template <typename... FArgs> std::tuple<FArgs...> getExpectedArgs(void(FArgs...)){};
```

**Comment by pvelesko** [comment]
Able to build CHIP-SPV and all the unit test (the ones that can't be built due to missing support in CHIP-SPV have been disabled)

**Comment by gargrahul** [comment]
@pvelesko There are many breaking changes in this PR, which are breaking build on AMD and Nvidia backends. Could you test your PR against AMD and Nvidia backends?

File: tests/catch/CMakeLists.txt [hunk 5]
```diff
@@ -46,32 +50,33 @@ message(STATUS "ROCM Path: ${ROCM_PATH}")
 
 
 if(UNIX)
-    set(CMAKE_CXX_COMPILER "${HIP_PATH}/bin/hipcc")
-    set(CMAKE_C_COMPILER "${HIP_PATH}/bin/hipcc")
-    set(HIPCONFIG_EXECUTABLE "${HIP_PATH}/bin/hipconfig")
-    execute_process(COMMAND perl ${HIPCONFIG_EXECUTABLE} --version
-                    OUTPUT_VARIABLE HIP_VERSION
-                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+    set(CMAKE_CXX_COMPILER "${HIP_PATH}/bin/hipcc.bin")
```
Comment (gargrahul) [inline tests/catch/CMakeLists.txt:53]: Currently usage of hipcc.bin is not enabled by default. You can use an env var HIP_USE_PERL_SCRIPTS=0 to enable hipcc.bin usage through hipcc.

File: tests/catch/CMakeLists.txt [hunk 6]
```diff
@@ -46,32 +50,33 @@ message(STATUS "ROCM Path: ${ROCM_PATH}")
...
-    # using cmake_path as it handles path correctly.
-    # Set both compilers else windows cmake complains of mismatch
-    cmake_path(SET CMAKE_CXX_COMPILER "${HIP_PATH}/bin/hipcc.bat")
-    cmake_path(SET CMAKE_C_COMPILER "${HIP_PATH}/bin/hipcc.bat")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --rocm-path=${ROCM_PATH}")
-    set(HIPCONFIG_EXECUTABLE "${HIP_PATH}/bin/hipconfig.bat")
-    execute_process(COMMAND ${HIPCONFIG_EXECUTABLE} --version
-                    OUTPUT_VARIABLE HIP_VERSION
-                    OUTPUT_STRIP_TRAILING_WHITESPACE)
+    # # using cmake_path as it handles path correctly.
```
Comment (gargrahul) [inline tests/catch/CMakeLists.txt:61]: On Windows, hipcc.bat is being used to invoke hipcc perl wrapper. Commenting of these lines is not correct.

File: tests/catch/unit/event/Unit_hipEventRecord.cc [hunk 7]
```diff
@@ -85,9 +85,10 @@ TEST_CASE("Unit_hipEventRecord") {
     // Record the start event
     HIP_CHECK(hipEventRecord(start, NULL));
 
-    HipTest::launchKernel<float>(HipTest::vectorADD<float>, blocks, 1, 0, 0,
-                                 static_cast<const float*>(A_d), static_cast<const float*>(B_d),
-                                 C_d, N);
+    // TODO compilation failure
```
Comment (gargrahul) [inline tests/catch/unit/event/Unit_hipEventRecord.cc:88]: Why are you commenting these? These calls are working fine on AMD and Nvidia backends.

File: tests/catch/unit/occupancy/hipOccupancyMaxPotentialBlockSize.cc [hunk 8]
```diff
@@ -69,6 +69,20 @@ TEST_CASE("Unit_hipOccupancyMaxPotentialBlockSize_rangeValidation") {
 
 }
 
+/*
```
Comment (gargrahul) [inline tests/catch/unit/occupancy/hipOccupancyMaxPotentialBlockSize.cc:72]: These lines shouldn't be commented. These are building fine on AMD and Nvidia backends. For SPIRV you could disable this test in cmake.

**Comment by gargrahul** [comment]
> Encountered the following error while building tests: @gargrahul
> 
> ```
> [ 67%] Building CXX object catch/catch_tests/unit/memory/CMakeFiles/MemoryTest.dir/hipHostMalloc.cc.o
> In file included from /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/unit/memory/hipHostMalloc.cc:32:
> In file included from /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_checkers.hh:24:
> /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:242:38: error: no matching function for call to 'getExpectedArgs'
>   using expectedArgsTuple = decltype(getExpectedArgs(f));
>                                      ^~~~~~~~~~~~~~~
> /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:267:3: note: in instantiation of function template specialization 'HipTest::validateArguments<void (*)(int *, int), int *, int>' requested here
>   validateArguments(kernel, packedArgs...);
>   ^
> /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/unit/memory/hipHostMalloc.cc:69:14: note: in instantiation of function template specialization 'HipTest::launchKernel<void (*)(int *, int), dim3, int *&, int>' requested here
>     HipTest::launchKernel(Set, dimGrid, dimBlock, 0, 0x0, ptr, -42);
>              ^
> /gpfs/jlse-fs0/users/pvelesko/CHIP-SPV/HIP/tests/catch/./include/hip_test_common.hh:228:51: note: candidate template ignored: deduced type 'void (*)(int *, int)' of 1st parameter does not match adjusted type 'void (*)(int *, int)' of argument [with FArgs = <int *, int>]
> template <typename... FArgs> std::tuple<FArgs...> getExpectedArgs(void(FArgs...)){};
> ```

We are not able to reproduce this error. Tests are being built on CI  used by this project.

File: tests/catch/CMakeLists.txt (hunk 5 above)
Comment (pvelesko) [reply tests/catch/CMakeLists.txt:53]: Even when using Perl-based hipcc, I use `hipcc ./file.cc ....` not `perl hipcc ./file.cc ...` I figured this is more universal?

File: tests/catch/unit/event/Unit_hipEventRecord.cc (hunk 7 above)
Comment (pvelesko) [reply tests/catch/unit/event/Unit_hipEventRecord.cc:88]: Due to compilation error related to template deduction.

File: tests/catch/unit/occupancy/hipOccupancyMaxPotentialBlockSize.cc (hunk 8 above)
Comment (pvelesko) [reply tests/catch/unit/occupancy/hipOccupancyMaxPotentialBlockSize.cc:72]: This is a work in progress - I'm was waiting for some feedback. 

> These lines shouldn't be commented. These are building fine on AMD and Nvidia backends

I will investigate why this is happening.

**Comment by pvelesko** [comment]
Closing in favor of a new branch branch

**Pull request closed without merging**